lib/Common/DataStructures/WeakReferenceDictionary.h

```diff
@@ -44,4 +44,670 @@ namespace JsUtil
+                            currentIndex = nextIndex;
+                        }
+                        else
+                        {
+                            lastIndex = currentIndex;
+                            currentIndex = entries[currentIndex].next;
+                        }
+                    }
+                }
+            }
+        }
+
+        void Initialize(int capacity)
+        {
+            // minimum capacity is 4
+            int initSize = max(capacity, 4);
+            int modIndex = UNKNOWN_MOD_INDEX;
+            uint initBucketCount = SizePolicy::GetBucketSize(initSize, &modIndex);
+            AssertMsg(initBucketCount > 0, "Size returned by policy should be greater than 0");
+
+            int* newBuckets = nullptr;
+            EntryType* newEntries = nullptr;
+            ValueType* newValues = nullptr;
+            Allocate(&newBuckets, &newEntries, &newValues, initBucketCount, initSize);
+
+            // Allocation can throw - assign only after allocation has succeeded.
+            this->buckets = newBuckets;
+            this->entries = newEntries;
+            this->values = newValues;
+            this->bucketCount = initBucketCount;
+            this->size = initSize;
+            this->modFunctionIndex = modIndex;
+            Assert(this->freeCount == 0);
+
+        }
+
+
+        template <InsertOperations op>
+        int Insert(const TKey& key, const TValue& value)
+        {
+            int * localBuckets = buckets;
+            if (localBuckets == nullptr)
+            {
+                Initialize(0);
+                localBuckets = buckets;
+            }
+
+#if DBG
+            // Always search and verify
+            const bool needSearch = true;
+#else
+            const bool needSearch = (op != Insert_Add);
+#endif
+            hash_t hashCode = GetHashCode(key);
+            uint targetBucket = this->GetBucket(hashCode);
+            if (needSearch)
+            {
+                EntryType * localEntries = entries;
+                for (int i = localBuckets[targetBucket]; i >= 0; i = localEntries[i].next)
+                {
+                    if (localEntries[i].template KeyEquals<Comparer<TKey>>(key, hashCode))
+                    {
+                        Assert(op != Insert_Add);
+                        if (op == Insert_Item)
+                        {
+                            values[i] = value;
+                            return i;
+                        }
+                        return -1;
+                    }
+                }
+            }
+
+            // Ideally we'd do cleanup only if weak references have been collected since the last resize
+            // but that would require us to use an additional field to store the last recycler cleanup id
+            // that we saw
+            // We can add that optimization later if we have to.
+            if (EntryType::SupportsCleanup() && freeCount == 0 && count == size)
```

Comment:
>EntryType::SupportsCleanup() [](start = 16, length = 28)

At a glance it looks like our entry type would say it doesn't support cleanup, but we actually can clean up. Maybe skip this check?

Reply:
or compile assert that it's true, since this dictionary is for a pretty specific use case?

---
In reply to: [187403074](https://github.com/Microsoft/ChakraCore/pull/5121#discussion_r187403074) [](ancestors = 187403074)

Reply:
This dictionary is inherently able to clean up, so I'll remove the check and update accordingly.